File: kotlin-node/karakum/src/jsMain/resources/plugins/convertOverriddenMethodDeclaration.js
import ts from "typescript";
import * as karakum from "../karakum.mjs";

const overriddenProps = {
    // stream
    "ReadableBase": {
        "read": "readOrNull"
    },

    // http2
    "Http2ServerRequest": {
        "read": "readOrNull"
    }
}

export default function (node, context, render) {
    if (
        ts.isMethodDeclaration(node)
        && ts.isClassDeclaration(node.parent)
        && node.parent.name
        && node.parent.name.text in overriddenProps
        && ts.isIdentifier(node.name)
        && node.name.text in overriddenProps[node.parent.name.text]
    ) {
        const name = karakum.escapeIdentifier(render(node.name))

        const typeParameters = node.typeParameters
            ?.map(typeParameter => render(typeParameter))
            ?.join(", ")

        const returnType = node.type && render(node.type)

        const overrideConfig = overriddenProps[node.parent.name.text][node.name.text]

        const hasAlias = overrideConfig !== null

        if (hasAlias) {
            let alias
            let test = () => true

            if (Array.isArray(overrideConfig)) {
                [alias, test] = overrideConfig
            } else {
                alias = overrideConfig
            }

            const inheritanceModifierService = context.lookupService(karakum.inheritanceModifierServiceKey.get())

            return karakum.convertParameterDeclarations(node, context, render, {
                strategy: "function",
                template: (parameters, signature) => {
                    if (!test(signature, node)) {
                        return `fun ${karakum.ifPresent(typeParameters, it => `<${it}> `)}${name}(${parameters})${karakum.ifPresent(returnType, it => `: ${it}`)}`
                    }

                    const inheritanceModifier = inheritanceModifierService?.resolveSignatureInheritanceModifier(node, signature, context)

                    const jsName = inheritanceModifier !== "override"
                        ? `@JsName("${name}")`
                        : ""

                    return `
${jsName}
${karakum.ifPresent(inheritanceModifier, it => `${it} `)}fun ${karakum.ifPresent(typeParameters, it => `<${it}> `)}${alias}(${parameters})${karakum.ifPresent(returnType, it => `: ${it}`)}
                    `
                }
            })
        }

        return karakum.convertParameterDeclarations(node, context, render, {
            strategy: "function",
            template: parameters => {
                return `override fun ${karakum.ifPresent(typeParameters, it => `<${it}> `)}${name}(${parameters})${karakum.ifPresent(returnType, it => `: ${it}`)}`
            }
        })
    }
    return null
}

File: kotlin-node/karakum/src/jsMain/kotlin/node/karakum/plugins/convertOverriddenMethodDeclaration.kt
package node.karakum.plugins

import node.karakum.util.nullable
import io.github.sgrishchenko.karakum.extension.createPlugin
import io.github.sgrishchenko.karakum.extension.ifPresent
import io.github.sgrishchenko.karakum.extension.plugins.InheritanceModifierService
import io.github.sgrishchenko.karakum.extension.plugins.ParameterDeclarationStrategy
import io.github.sgrishchenko.karakum.extension.plugins.ParameterDeclarationsConfiguration
import io.github.sgrishchenko.karakum.extension.plugins.convertParameterDeclarations
import io.github.sgrishchenko.karakum.extension.plugins.inheritanceModifierServiceKey
import io.github.sgrishchenko.karakum.util.escapeIdentifier
import io.github.sgrishchenko.karakum.util.getParentOrNull
import typescript.asArray
import typescript.isClassDeclaration
import typescript.isIdentifier
import typescript.isMethodDeclaration

private val overriddenMethods = mapOf(
    // stream
    "Readable" to mapOf(
        "read" to "readOrNull"
    ),

    // http2
    "Http2ServerRequest" to mapOf(
        "read" to "readOrNull"
    ),
)

val convertOverriddenMethodDeclaration = createPlugin { node, context, render ->
    nullable {
        ensure(isMethodDeclaration(node))

        val classNode = ensureNotNull(node.getParentOrNull())
        ensure(isClassDeclaration(classNode))

        val className = ensureNotNull(classNode.name)
        val classOverrides = ensureNotNull(overriddenMethods[className.text])

        val methodName = node.name
        ensure(isIdentifier(methodName))
        val methodOverride = ensureNotNull(classOverrides[methodName.text])

        val name = escapeIdentifier(render(node.name))

        val typeParameters = node.typeParameters?.asArray()
            ?.joinToString(", ") { render(it) }

        val returnType = node.type?.let {render(it) }

        val inheritanceModifierService =
            context.lookupService<InheritanceModifierService>(inheritanceModifierServiceKey)

        convertParameterDeclarations(node, context, render, ParameterDeclarationsConfiguration(
            strategy = ParameterDeclarationStrategy.function,
            template = { parameters, signature ->

            val inheritanceModifier = inheritanceModifierService?.resolveSignatureInheritanceModifier(node, signature, context)

            val jsName = if (inheritanceModifier != "override") "@JsName(\"$name\")" else ""

            """
                $jsName
                ${ifPresent(inheritanceModifier) { "$it " }}fun ${ifPresent(typeParameters) { "<$it> " }}${methodOverride}(${parameters})${ifPresent(returnType) { ": $it" }}
            """.trimIndent()
        }))
    }
}
